### ISSUE TITLE
It's possible to crop an live video and make predictions in this cropped frame?

### ISSUE THREAD

**Issue by ViniciusRosa1** [ultralytics/ultralytics#16464]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am working on a project using a local camera with RTSP, but I'm facing an FPS issue. So, I'm trying to detect the person, crop the bounding box of the detected person, and make EPI predictions on the cropped person image. The idea is to save this data in a CSV to track the objects detected on the person. The following code demonstrates a bit of what I'm trying to do. I'm open to any kind of ideas or help!

````python
from datetime import timedelta
import json
import pandas as pd
import torch
import numpy as np
import cv2
import mediapipe as mp
from time import time
from ultralytics import YOLO
from videostream import VideoStream
import os
import supervision as sv

classes = ['Hardhat', 'Mask', 'NO-Hardhat', 'NO-Mask', 'NO-Safety Vest', 'Person', 'Safety Cone', 'Safety Vest', 'machinery', 'vehicle']
ip_url = "rtsp://root:admin@ip.address/axis-media/media.amp?adjustablelivestream=1"

class ObjectDetection:
    def __init__(self, capture_index):
       
        self.capture_index = capture_index
        self.video_stream = VideoStream(source=capture_index).start()
        self.device = 'cuda' if torch.cuda.is_available() else 'cpu'
        print("Using Device: ", self.device)
        self.model = self.load_model()
        self.CLASS_NAMES_DICT = self.model.model.names
        self.box_annotator = sv.BoxAnnotator(sv.ColorPalette.default(), thickness=3, text_thickness=3, text_scale=1.5)
        
        self.frames_generator = sv.get_video_frames_generator(ip_url)

    

    def load_model(self):
       
        model = YOLO("../best.pt")  # load a pretrained YOLOv8n model
        model.fuse()
    
        return model
    
    def predict(self, frame):
        results = self.model(frame, save_crop= True)
        return results

    def crop_person(self, frame, results):
        for result in results:
            boxes = result.boxes  
            for i in range(len(boxes)):  
                class_id = int(boxes.cls[i].item())  
                if class_id == 5:  
                    xyxy = boxes.xyxy[i].cpu().numpy()  # Obtém as coordenadas da bounding box
                    x_min, y_min, x_max, y_max = map(int, xyxy)  # Converte as coordenadas para inteiros
                    cropped_image = frame[y_min:y_max, x_min:x_max]  # Recorta a imagem no frame
                    return cropped_image  # Retorna a imagem recortada
        return None  # Retorna None se nenhuma pessoa for encontrada

    
    def detect_epis(self, cropped_frame):
        if cropped_frame is not None:
            results = self.predict(cropped_frame)
            self.plot_bboxes(results, cropped_frame)
            
            
    def plot_bboxes(self, results, frame):
        xyxys = []
        confidences = []
        class_ids = []
         # Extract detections for person class
        for result in results:
            boxes = result.boxes.cpu().numpy()
            for i in range(len(boxes)):
                class_id = int(boxes.cls[i])
                conf = boxes.conf[i].cpu().numpy()
                xyxy = boxes.xyxy[i].cpu().numpy()
                label = f"{self.CLASS_NAMES_DICT[class_id]} {conf:.2f}"
      
                    
                # Crop the detected region from the frame
                x_min, y_min, x_max, y_max = map(int, xyxy)  # Convert to integer
                cv2.rectangle(frame, (x_min, y_min), (x_max, y_max), (0, 255, 0), 2)
                cv2.putText(frame, label, (x_min, y_min - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 0, 0), 2)

        
        # Setup detections for visualization
        #detections = sv.Detections(
                    # xyxy=results[0].boxes.xyxy.cpu().numpy(),
                    # confidence=results[0].boxes.conf.cpu().numpy(),
                    # class_id=results[0].boxes.cls.cpu().numpy().astype(int),
                    # )
 
        # Format custom labels
        # self.labels = [f"{self.CLASS_NAMES_DICT[class_id]} {confidence:0.2f}"
        # for _, confidence, class_id, tracker_id
        # in detections]
        
        # # Annotate and display frame
        # frame = self.box_annotator.annotate(scene=frame, detections=detections, labels=self.labels)
        
        # return frame

    def saveDetectionsToCSV(results, model, outputFile='detection_results.csv'):
        data = []
        for imgPath, result in zip(results.files, results):
            for box in result.boxes:
                conf = box.conf.item()  # Confiança da detecção
                cls = int(box.cls.item())  # Classe detectada
                data.append({
                    "Image/File Name": imgPath,
                    "Detected Class": model.names[cls],  # Nome da classe detectada
                    "Confidence": conf,  # Nível de confiança
                    "Bounding Box": box.xyxy.tolist()  # Coordenadas da bounding box
                })
        
        # Convertendo para DataFrame e salvando como CSV
        df = pd.DataFrame(data)
        df.to_csv(outputFile, index=False)
        
        
        
    def __call__(self):

        cap = cv2.VideoCapture(self.capture_index)
        assert cap.isOpened()
        cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
        cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)
        cap.set(cv2.CAP_PROP_BUFFERSIZE,2)
        
        # Maximum number of frames to be analyzed
        cap.get(cv2.CAP_PROP_FPS)
        
        
        
        # mp_pose = mp.solutions.pose
        # pose_video = mp_pose.Pose(static_image_mode=False, model_complexity=2, min_tracking_confidence=0.5, min_detection_confidence=0.5)
        # mp_drawing = mp.solutions.drawing_utils
        
        
        while True:
            start_time = time()
            frame = self.video_stream.read()
            
                
            results = self.predict(frame)
            cropped_person  = self.crop_person(results, frame)
            self.detect_epis(cropped_person)
            
            
            end_time = time()
            fps = 1/np.round(end_time - start_time, 2)
            
            cv2.putText(frame, f'FPS: {int(fps)}', (20,70), cv2.FONT_HERSHEY_SIMPLEX, 1.5, (0,255,0), 2)
                
            cv2.imshow('YOLOv8 Detection', frame)
 
            if cv2.waitKey(5) & 0xFF == ord('q'):
                self.video_stream.stop()
                break
        cv2.destroyAllWindows()
        
        
    
detector = ObjectDetection(capture_index=ip_url)
detector()
````

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ViniciusRosa1, thank you for your interest in Ultralytics 🚀! We suggest checking out the [Docs](https://docs.ultralytics.com) for resources that might help, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples.

For your custom training ❓, be sure to share detailed information, including dataset examples and training logs, and confirm you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) for faster assistance.

Join the Ultralytics community in any channel that suits you. For real-time interaction, join us on [Discord](https://ultralytics.com/discord) 🎧. Prefer detailed discussions? Head to [Discourse](https://community.ultralytics.com) or explore our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you're on the latest version 🚀 of the `ultralytics` package, along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [Python>=3.8](https://www.python.org/) environment and [PyTorch>=1.8](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLOv8 is optimized for the following verified environments (with all dependencies like [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon AWS**: Check our [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: Follow our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing. CI tests ensure proper function of all YOLOv8 [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and with each commit.

This is an automated response, but an Ultralytics engineer will be with you soon. 😊

**Comment by Y-T-G** [comment]
Well, the FPS will of course be low because you're writing to a file each time. It's an I/O operation which would severely handicap the performance. You should only be writing every 30 or 60 frames at once and be using threading so that you don't block the main thread.
